streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -955,20 +969,28 @@ boolean maybeCommit() {
                     taskManager.activeTaskIds(), taskManager.standbyTaskIds(), now - lastCommitMs, commitTimeMs);
             }
 
-            committed = taskManager.commitAll();
+            committed = taskManager.commitAllActive();
             if (committed > 0) {
+                if (eosBetaEnabled) {
+                    producer.commitTransaction();
+                    producer.beginTransaction();
```

Comment:
\cc @vvcephei 

With KIP-447 we commit all tasks at once in the thread -- we could actually do the same with non-eos and consumer-commits (atm, we do a commit roundtrip per task what seems to be inefficient). Do you think your refactoring work would allow to restructure the code accordingly?

Reply:
@mjsax , just to clarify, are you asking whether https://github.com/guozhangwang/kafka/pull/6 would conflict in some way with restructuring the commit logic? I don't believe so. I'm only changing how the tasks' lifecycles are managed, so it ought to be encapsulated from the thread's perspective.

Reply:
I am also raising the question, if we should refactor how we do non-EOS offset commits. Might be an independent PR and could be part of your refactoring efforts.

Reply:
I think moving forward if we are always going to commit all tasks owned by a thread anyways, we should consider separating the logic of 

1) flushing stores, writing checkpoint files
2) calling consumer / producer APIs for commit (or sendOffsets)

For 1) it is per-task, while 2) is per-thread. In the refactoring atm we move all of 2) into the RecordCollector which is per-task, but maybe in the future we can make the RecordCollector per-thread shared by tasks to make step 2) "global".